app/src/main/java/de/tum/in/tumcampusapp/component/tumui/roomfinder/RoomFinderDetailsActivity.kt

```diff
@@ -0,0 +1,300 @@
+                getRoomTimetable()
+                invalidateOptionsMenu()
+                return true
+            }
+            R.id.action_directions -> {
+                loadGeo()
+                return true
+            }
+            R.id.action_switch_map -> {
+                showMapSwitch()
+                return true
+            }
+            else -> return super.onOptionsItemSelected(item)
+        }
+    }
+
+    override fun onBackPressed() {
+        // Remove fragment with room timetable if present and show map again
+        if (fragment != null) {
+            getRoomTimetable()
+            invalidateOptionsMenu()
+            return
+        }
+
+        super.onBackPressed()
+    }
+
+    private fun getRoomTimetable() {
+        val ft = supportFragmentManager.beginTransaction()
+        // Remove if fragment is already present
+        if (fragment != null) {
+            ft.replace(R.id.fragment_container, mImageFragment)
+            ft.commit()
+            fragment = null
+            return
+        }
+
+        val roomApiCode = room.room_id
+        fragment = WeekViewFragment.newInstance(roomApiCode)
+        ft.replace(R.id.fragment_container, fragment!!)
+        ft.commit()
+    }
+
+    private fun showMapSwitch() {
+        val list = arrayOfNulls<CharSequence>(mapsList.size)
+        var curPos = 0
+        for (index in mapsList.indices) {
+            list[index] = mapsList[index]
+                    .description
+            if (mapsList[index].map_id == mapId) {
+                curPos = index
+            }
+        }
+        AlertDialog.Builder(this).setSingleChoiceItems(list, curPos, this)
+                .show()
+    }
+
+    override fun onClick(dialog: DialogInterface, whichButton: Int) {
+        dialog.dismiss()
+        val selectedPosition = (dialog as AlertDialog).listView.checkedItemPosition
+        mapId = mapsList[selectedPosition].map_id
+        startLoading()
+    }
+
+    override fun onLoadInBackground(vararg arg: Void): String? {
+        val encodedArchId = ApiHelper.encodeUrl(room.arch_id)
+
+        return if (mapId.isEmpty()) {
+            Const.URL_DEFAULT_MAP_IMAGE + encodedArchId
+        } else {
+            Const.URL_MAP_IMAGE + encodedArchId + '/'.toString() + ApiHelper.encodeUrl(mapId)
+        }
+    }
+
+    override fun onLoadFinished(url: String?) {
+        mImageFragment.loadImage(url!!, object : ImageViewTouchFragment.ImageLoadingListener {
+            override fun onImageLoadingError() {
+                showImageLoadingError()
+            }
+        })
+
+        supportActionBar?.title = room.info
+        supportActionBar?.subtitle = room.formattedAddress
+
+        showLoadingEnded()
+        loadMapList()
+    }
+
+    private fun loadMapList() {
+        showLoadingStart()
+
+        mRoomFinderMapsCall = TUMCabeClient.getInstance(this).fetchAvailableMaps(room.arch_id)
+        mRoomFinderMapsCall?.enqueue(object : Callback<List<RoomFinderMap>> {
```

Comment:
If you feel up for it, you could try to use RxJava here instead. Should make for less code and easier cancellation inn `onDestroy()` via a `CompositeDisposable`. Not a must, though.